docs/developers/API_Reference.md

```diff
@@ -0,0 +1,317 @@
+
+To create a custom map object:
+
+1. Inherit from `CGObjectInstance`
+2. Override required virtual methods
+3. Register the object type in `RegisterTypes.h`
+
+```cpp
+class MyCustomObject : public CGObjectInstance
+{
+public:
+    void onHeroVisit(const CGHeroInstance * h) const override;
+    void initObj(vstd::RNG & rand) override;
+    
+    template<typename Handler>
+    void serialize(Handler &h, const int version)
+    {
+        h & static_cast<CGObjectInstance&>(*this);
+        // serialize custom data
+    }
+};
+```
+
+### Object Interaction
+
+```cpp
+// Visit callbacks
+virtual void onHeroVisit(const CGHeroInstance * h) const;
+virtual void onHeroLeave(const CGHeroInstance * h) const;
+
+// Battle integration
+virtual void battleFinished(const CGHeroInstance *hero, const BattleResult &result) const;
+```
+
+## Bonus System API
+
+The bonus system is VCMI's core mechanism for handling all stat modifications and special abilities.
+
+### Basic Bonus Usage
+
+```cpp
+// Create a bonus
+std::shared_ptr<Bonus> bonus = std::make_shared<Bonus>();
+bonus->type = Bonus::PRIMARY_SKILL;
+bonus->subtype = PrimarySkill::ATTACK;
+bonus->val = 5; // +5 attack
+
+// Apply to object
+bonusSystemNode->addNewBonus(bonus);
+```
+
+### Bonus Types
+
+- `PRIMARY_SKILL` - Hero primary skills (Attack, Defense, Spell Power, Knowledge)
+- `SECONDARY_SKILL_PREMY` - Secondary skill bonuses
+- `STACKS_SPEED` - Stack speed modifications
+- `MAGIC_RESISTANCE` - Magic resistance
+- And many more...
+
+For complete reference, see [Bonus System](Bonus_System.md) documentation.
+
+## Scripting API
+
+### Lua Integration
+
+VCMI supports Lua scripting for game logic extension.
+
+#### Basic Script Structure
+
+```lua
+local function onVisit(player, object)
+    -- Script logic here
+    return true
+end
+
+return {
+    onVisit = onVisit
+}
+```
+
+#### Available APIs
+
+```lua
+-- Game state access
+local hero = game:getHero(heroId)
+local town = game:getTown(townId)
+
+-- Object manipulation
+object:addReward(reward)
+object:setProperty(property, value)
+
+-- Player interaction
+player:addResource(resourceType, amount)
+player:giveArtifact(artifactId)
+```
+
+For detailed scripting reference, see [Lua Scripting System](Lua_Scripting_System.md).
+
+## Networking API
+
+### Network Packets
+
+All communication between client and server uses serializable packet classes inheriting from `CPack`.
+
+#### Creating Custom Packets
+
+```cpp
+class MyCustomPack : public CPack
+{
+public:
+    ObjectInstanceID objectId;
+    si32 customData;
+    
+    void applyGs(CGameState *gs); // Server-side application
+    
+    template<typename Handler>
+    void serialize(Handler &h, const int version)
+    {
+        h & objectId;
+        h & customData;
+    }
+};
+```
+
+#### Sending Packets
+
+```cpp
+// From client to server
+sendRequest(new MyCustomPack(objectId, data));
+
+// From server to client
+sendToAllClients(new MyCustomPack(objectId, data));
```

Comment:
These methods either pass pack by reference or by smart pointer, not by raw pointer